### PR TITLE
[PP-48] Check user vote through the game progress

### PR DESCRIPTION
**Pull request by MakFill**
Task: https://andrei-lysiuk.atlassian.net/browse/PP-48
Integration vote system with BE

### CODE REVIEW

File: src/components/CommonArea/CardNest/CardNest.tsx [hunk 1]
```diff
@@ -1,40 +1,47 @@
...
+        className,
+        {
+          [styles.spectator]: userRole === 'spectator',
+          [styles.showCard]: roomState === StateRoomTitle.showCards,
+        },
       )}
-  </section>
-);
+    >
+      {userRole !== 'spectator'
```
Comment (mrakofff) [inline src/components/CommonArea/CardNest/CardNest.tsx]: do user still have role if he is spectator?

File: src/components/CommonArea/Table/Table.tsx [hunk 2]
```diff
@@ -1,44 +1,47 @@
...
-export const Table: React.FC<ITableProps> = ({
-  selectedCardValue,
-  isCardOpened,
-  setIsCardIsVisible,
-  setSelectedCardValue,
-}) => {
   function showCards() {
-    setIsCardIsVisible(true);
+    if (activeTask) {
+      socketService.emit(SocketEvent.RoomShow, { id: activeTask.id });
```
Comment (mrakofff) [inline src/components/CommonArea/Table/Table.tsx]: why payload is object but not a string?

File: src/components/Controls/Controls.tsx [hunk 3]
```diff
@@ -1,48 +1,102 @@
...
-      </SecondaryButton>
-    </section>
+    <>
+      { role
+     && (
+       <section className={styles.groupBtn}>
+         <PrimaryButton
+           className={classNames(styles.singleBtn,
+             {
+               [styles.disable]: !activeTask,
```
Comment (mrakofff) [inline src/components/Controls/Controls.tsx]: you could add styles for `:disabled` pseudo-class

File: src/store/reducers/roomStateReducer.ts [hunk 4]
```diff
@@ -1,16 +1,18 @@
 import { createReducer } from '@reduxjs/toolkit';
 import { roomStateActions, userActions } from '@/store/actions';
+import { StateRoomTitle } from '@/utils/enums';
 
 const initialState = {
   roomTitle: '',
   voteSystem: '',
   dealerRights: '',
   roomId: '',
+  roomState: StateRoomTitle.reset.toString(),
```
Comment (mrakofff) [inline src/store/reducers/roomStateReducer.ts]: why `.toString`?

File: src/store/actions/roomStateAction.ts [hunk 5]
```diff
@@ -4,24 +4,26 @@ import { IRoomState, IRoomFromBE } from '@/utils/interfaces';
 import { history } from '@/utils/history';
 import { getRoomIdByUrl } from '@/helpers';
 import { navMap } from '@/utils/NavMap';
+import { RoomStateModel } from '@/models';
 
 interface IRoomSettings {
   roomTitle: string;
   voteSystem?: string;
-  dealerRights?: string
+  dealerRights?: string;
```
Comment (mrakofff) [inline src/store/actions/roomStateAction.ts:12]: do we use this field?

File: src/containers/SendMessageForm/SendMessageForm.tsx [hunk 6]
```diff
@@ -37,8 +38,11 @@ export const SendMessageForm: React.FC = () => {
         onChange={sendMessageHandler}
       />
       <PrimaryButton
-        className={styles.submit}
+        className={classNames(styles.submit, {
```
Comment (mrakofff) [inline src/containers/SendMessageForm/SendMessageForm.tsx]: why don't use `disabled` props?

File: src/components/User/User.tsx [hunk 7]
```diff
@@ -22,13 +22,13 @@ export const User: FC<IUserProps> = ({ user }) => {
...
         <div className={styles.name}>
-          {user.name}
+          {name}
         </div>
         <div className={styles.kick} />
       </div>
       <CardNest
-        selectedCardValue={user.score || ''}
-        userRole={user.role}
+        score={user.score || ''}
```
Comment (mrakofff) [inline src/components/User/User.tsx]: you could destcruture from `user` on 12 row, and set default value in `CardNest`

File: src/components/Task/Task.tsx [hunk 8]
```diff
@@ -50,26 +51,30 @@ export const Task: React.FC<ITaskProps> = ({ task }) => {
...
+            <div className={styles.kick} />
+          </Button>
+        )}
       </div>
       <p className={styles.task__title}>{task.title}</p>
       <div className={styles.task__bottom}>
         <PrimaryButton
           className={classNames(styles.vote_btn, {
-            [styles.vote_btn__disabled]: task.isActive,
+            [styles.vote_btn__disabled]: task.isActive || role !== 'admin',
```
Comment (mrakofff) [inline src/components/Task/Task.tsx]: `disabled` prop?

File: src/components/Task/Task.tsx [hunk 9]
```diff
@@ -1,19 +1,21 @@
...
 import { cards } from '@/mocks/cards';
+import { userSelectors } from '@/store/selectors';
 
 interface ITaskProps {
   task: ITask;
 }
 
 export const Task: React.FC<ITaskProps> = ({ task }) => {
-  const [score, setScore] = useState(task.score || '?');
+  const role = useSelector(userSelectors.role)?.role;
```
Comment (mrakofff) [inline src/components/Task/Task.tsx]: `role.role`?

File: src/components/CommonArea/Table/Table.tsx [hunk 10]
```diff
@@ -1,44 +1,47 @@
...
+        && roomState !== StateRoomTitle.showCards
+        && score
         && <PrimaryButton onClick={() => showCards()}>Show cards</PrimaryButton>
       }
       {
-        isCardOpened
-        && <SecondaryButton onClick={() => removeChosenCard()}>New voting</SecondaryButton>
+        role === 'admin'
+        && roomState === StateRoomTitle.showCards
+          && <SecondaryButton onClick={() => gameFinish()}>Reset</SecondaryButton>
```
Comment (mrakofff) [inline src/components/CommonArea/Table/Table.tsx]: remove redundant arrow function and use as callback

File: src/store/selectors/userSelectors.ts [hunk 11]
```diff
@@ -7,5 +7,6 @@ export const userSelectors = {
   userId: createSelector(rootStateSelector, (state) => state.userId),
   name: createSelector(rootStateSelector, (state) => state.name),
   score: createSelector(rootStateSelector, (state) => state.score),
-  role: createSelector(rootStateSelector, (state) => state.role),
+  role: createSelector((state: RootState) => state,
+    (state) => state.members.find((member) => member.userId === state.user.userId)),
```
Comment (mrakofff) [inline src/store/selectors/userSelectors.ts]: - why don't you store userRole in userReducer?
- why `roleSelector` returns `IUser` but not `IUser['role']`

**Review by mrakofff — CHANGES_REQUESTED** [review]

File: src/components/CommonArea/Table/Table.tsx (hunk 2 above)
Comment (MakFill) [reply src/components/CommonArea/Table/Table.tsx]: refactored

File: src/components/Controls/Controls.tsx (hunk 3 above)
Comment (MakFill) [reply src/components/Controls/Controls.tsx]: refactored

File: src/store/reducers/roomStateReducer.ts (hunk 4 above)
Comment (MakFill) [reply src/store/reducers/roomStateReducer.ts]: fixed

File: src/store/actions/roomStateAction.ts (hunk 5 above)
Comment (MakFill) [reply src/store/actions/roomStateAction.ts:12]: no, removed

File: src/containers/SendMessageForm/SendMessageForm.tsx (hunk 6 above)
Comment (MakFill) [reply src/containers/SendMessageForm/SendMessageForm.tsx]: refactored

File: src/components/User/User.tsx (hunk 7 above)
Comment (MakFill) [reply src/components/User/User.tsx]: refactored

File: src/components/Task/Task.tsx (hunk 8 above)
Comment (MakFill) [reply src/components/Task/Task.tsx]: refactored

File: src/components/CommonArea/Table/Table.tsx (hunk 10 above)
Comment (MakFill) [reply src/components/CommonArea/Table/Table.tsx]: refactored

**Review by andreikaptsiuh — APPROVED** [review]

File: src/components/Task/Task.tsx (hunk 9 above)
Comment (MakFill) [reply src/components/Task/Task.tsx]: fixed

File: src/store/selectors/userSelectors.ts (hunk 11 above)
Comment (MakFill) [reply src/store/selectors/userSelectors.ts]: refactored

File: src/components/CommonArea/CardNest/CardNest.tsx (hunk 1 above)
Comment (MakFill) [reply src/components/CommonArea/CardNest/CardNest.tsx]: removed as outdated

File: src/components/Controls/Controls.tsx [hunk 12]
```diff
@@ -1,48 +1,100 @@
...
-      <SecondaryButton
-        className={styles.singleBtn}
-        onClick={handlerStop}
-      >
-        {NAMES_BTN.stop}
-      </SecondaryButton>
-    </section>
+    <>
+      { role
+     && (
```
Comment (mrakofff) [inline src/components/Controls/Controls.tsx:74]: add condition before return like if(!role) {return null }

File: src/containers/SendMessageForm/SendMessageForm.tsx [hunk 13]
```diff
@@ -39,6 +39,7 @@ export const SendMessageForm: React.FC = () => {
       <PrimaryButton
         className={styles.submit}
         type="submit"
+        disabled={text.trim().length === 0}
```
Comment (mrakofff) [inline src/containers/SendMessageForm/SendMessageForm.tsx:42]: `!text.trim().length`

File: src/store/reducers/roomStateReducer.ts [hunk 14]
```diff
@@ -4,16 +4,16 @@ import { roomStateActions, userActions } from '@/store/actions';
...
 
 export const roomState = createReducer(initialState, (builder) => {
   builder
-    .addCase(roomStateActions.setRoomState, (state, action) => ({ ...state, ...action.payload }))
-    .addCase(roomStateActions.getRoomByUrl.fulfilled, (state, action) => {
-      state.roomId = action.payload;
-    })
+    .addCase(roomStateActions.setRoomState, (state, action) => ({ ...state, roomState: action.payload }))
+    .addCase(roomStateActions.getRoomByUrl.fulfilled, (state, action) => ({
+      ...state, roomTitle: action.payload.title, roomId: action.payload.id, roomState: action.payload.state,
```
Comment (mrakofff) [inline src/store/reducers/roomStateReducer.ts:16]: you could mutate and destructure action

File: src/components/Button/Button.module.scss [hunk 15]
```diff
@@ -29,6 +29,15 @@
...
     background-color: $second-color-light;
   }
+
+  
+  &:disabled {
+    background-color: $shadow-grey
+  }
+
+  &:disabled:hover {
+    background-color: $shadow-grey
```
Comment (mrakofff) [inline src/components/Button/Button.module.scss:39]: you could join with previous selector with `,`

File: src/components/CreateTaskPanel/CreateTaskPanel.tsx [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+  };
+
+  return (
+    <div className={styles.createTaskPanel}>
+      {role === 'admin'
+      && (
+        <PrimaryButton
+          className={classNames(
+            styles.taskEditor__btn, {
+              [styles.taskEditor__btn_close]: isTaskEditorOpen === true,
```
Comment (mrakofff) [inline src/components/CreateTaskPanel/CreateTaskPanel.tsx:44]: `[styles.taskEditor__btn_close]: isTaskEditorOpen`

File: src/components/InviteBtn/InviteBtn.tsx [hunk 17]
```diff
@@ -0,0 +1,34 @@
+import React, { useState } from 'react';
+import { CopyToClipboard } from 'react-copy-to-clipboard';
+import styles from './InviteBtn.module.scss';
+
+export const InviteBtn: React.FC = () => {
+  const [isVisible, setIsVisible] = useState(false);
+  const url = window.location.href;
```
Comment (mrakofff) [inline src/components/InviteBtn/InviteBtn.tsx:7]: could you use `useLocation` hook?

**Review by mrakofff — CHANGES_REQUESTED** [review]

**Pull request merged**